constantine-go/constantine.go

```diff
@@ -357,135 +390,136 @@ func (sig1 EthBlsSignature) AreEqual(sig2 EthBlsSignature) bool {
 	return bool(status)
 }
 
-func (sec *EthBlsSecKey) Validate() (bool, error) {
+func (sec *EthBlsSecKey) Validate() (err error) {
 	status := C.ctt_eth_bls_validate_seckey((*C.ctt_eth_bls_seckey)(sec))
 	if status != C.cttCodecScalar_Success {
-		err := errors.New(
+		err = errors.New(
 			C.GoString(C.ctt_codec_scalar_status_to_string(status)),
 		)
-		return false, err
+		return err
 	}
-	return true, nil
+	return nil
 }
 
-func (pub *EthBlsPubKey) Validate() (bool, error) {
+func (pub *EthBlsPubKey) Validate() (err error) {
 	status := C.ctt_eth_bls_validate_pubkey((*C.ctt_eth_bls_pubkey)(pub))
 	if status != C.cttCodecEcc_Success {
-		err := errors.New(
+		err = errors.New(
 			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
 		)
-		return false, err
+		return err
 	}
-	return true, nil
+	return nil
 }
 
-func (sig *EthBlsSignature) Validate() (bool, error) {
+func (sig *EthBlsSignature) Validate() (err error) {
 	status := C.ctt_eth_bls_validate_signature((*C.ctt_eth_bls_signature)(sig))
 	if status != C.cttCodecEcc_Success {
 		err := errors.New(
 			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
 		)
-		return false, err
+		return err
 	}
-	return true, nil
+	return nil
 }
 
-func (sec *EthBlsSecKey) Serialize(dst *[32]byte) (bool, error) {
-	status := C.ctt_eth_bls_serialize_seckey((*C.byte)(unsafe.Pointer(dst)),
+func (sec *EthBlsSecKey) Serialize() (dst Bytes32, err error) {
+	status := C.ctt_eth_bls_serialize_seckey((*C.byte)(unsafe.Pointer(&dst)),
 		(*C.ctt_eth_bls_seckey)(sec),
 	)
 	if status != C.cttCodecScalar_Success {
 		err := errors.New(
 			C.GoString(C.ctt_codec_scalar_status_to_string(status)),
 		)
-		return false, err
+		return dst, err
 	}
-	return true, nil
+	return dst, nil
 }
 
-func (pub *EthBlsPubKey) SerializeCompressed(dst *[48]byte) (bool, error) {
-	status := C.ctt_eth_bls_serialize_pubkey_compressed((*C.byte)(unsafe.Pointer(dst)),
+func (pub *EthBlsPubKey) SerializeCompressed() (dst Bytes48, err error) {
+	status := C.ctt_eth_bls_serialize_pubkey_compressed((*C.byte)(unsafe.Pointer(&dst)),
 		(*C.ctt_eth_bls_pubkey)(pub),
 	)
 	if status != C.cttCodecEcc_Success {
 		err := errors.New(
 			C.GoString(C.ctt_codec_ecc_status_to_string(status)),
 		)
-		return false, err
+		return dst, err
```

Comment:
```suggestion
		return _, err
```
I think we can return a default value here, I don't remember what Nim does there.

Reply:
Or maybe dst is already default init. Anyway we should assume that people check the error.

Reply:
> I think we can return a default value here, I don't remember what Nim does there.

Returning `_` is not valid in Go. We can either do `TheType{}` to explicitly zero mem or just rely on the fact that all values we return are stack allocated and thus zeroed anyway. 

On the Nim side (in the actual functions that implement the logic, not the module that generates the C API) it seems like in some cases we use `setZero` e.g. here:

https://github.com/mratsim/constantine/blob/master/constantine/serialization/codecs_bls12_381.nim#L109-L111

and in other cases we simply don't touch the value we return, e.g. here:

https://github.com/mratsim/constantine/blob/master/constantine/serialization/codecs_bls12_381.nim#L148-L149